### ISSUE TITLE
[Bug] Concat followed by ConstantOfShape produces node name resolution error

### ISSUE THREAD

**Issue by wyessen** [pytorch/glow#3249]
`loadConcat` produces a node that is not of type `Constant` (it is rather of type `Concat` which is a subclass of `Storage`). On the other hand, `loadConstantOfShape` expects a `Constant` node as input. So `Concat` followed by `ConstantOfShape` produces an error such as the following:

```
../../../lib/Importer/ProtobufLoader.cpp:79 message: could not find constant with name 43
```

The relevant part from an ONNX graph is:

```
 %40 : Tensor = onnx::Unsqueeze[axes=[0]](%38)
  %41 : Tensor = onnx::Unsqueeze[axes=[0]](%37)
  %42 : Tensor = onnx::Unsqueeze[axes=[0]](%39)
  %43 : Tensor = onnx::Concat[axis=0](%40, %41, %42)
  %44 : Float(1, 1, 128) = onnx::ConstantOfShape[value={0}](%43)
```

**Comment by ksaurabh-cadence** [comment]
I am not sure what the complete graph looks like. What about %38,%37 and %39, are they constants/shapes? Have you tried running running with const-fold-ops option?

**Comment by wyessen** [comment]
No they are not. They all depend on the input.

**Comment by wyessen** [comment]
Alright, so with that option, I get:

```
E0718 10:21:34.344532 268948928 Error.h:266] Converting error to boolean: location: ../../../include/glow/Importer/ProtobufLoader.h:223 message: No constant found
E0718 10:21:34.345816 268948928 Error.h:266] Converting error to boolean: location: ../../../include/glow/Importer/ProtobufLoader.h:223 message: No constant found
E0718 10:21:34.345906 268948928 Error.h:266] Converting error to boolean: location: ../../../include/glow/Importer/ProtobufLoader.h:223 message: No constant found
Encountered an error, exiting.
location: ../../../lib/Importer/ProtobufLoader.cpp:79 message: could not find constant with name 43
```

**Comment by wyessen** [comment]
The problem is the implementation of `loadConstantOfShape`, as it assumes that the input should be constant. There is no reason why the input should be constant, and unless I'm missing something, there is no reason to _require_ the input to be constant (after all, we're only grabbing the input's dimensions and are creating a constant tensor with those dimensions).

**Comment by wyessen** [comment]
Basically, what we need is a common `Node` subclass that carries a payload, like `Constant` does. Now, `Constant` is subclassed from `Storage`, and `Storage` is subclassed from `Node.` In this particular case, however, a `ConcatNode` is coming in, and that node is subclassed from `Node`. So there is no common interface between `Constant` and `ConcatNode`, other than `Node`, but `Node` does not carry a payload.

In other words, whenever a "generic" node with a payload is assumed, it has got to be a subclass of `Storage`.

**Comment by jfix71** [comment]
>They all depend on the input.

Glow doesn't currently support dynamic shapes. In this model, we cannot determine the shape of the Splat for the Constant based on the dynamic inputs of the model. The shape has to be statically known. This is why we need the input to be Constant.

**Comment by wyessen** [comment]
@jfix71 Well, technically we can iterate over the tensor elements of a `ConcatNode`, can we not? If so, then we should be able to create `ConstantOfShape` from a `ConcatNode`. In other words, as long as we're able to get a tensor payload and that tensor is of shape `1D`, and its elements are of type `Int64`, we should be able to take that as input to `loadConstantOfShape`. Or am I missing something?

**Comment by ayermolo** [comment]
This assumes we know payload of the tensor at compile time, which I don't think is usually the case...

**Comment by wyessen** [comment]
Well, this effectively means that some graphs produced by `torch.onnx` for some models (and not the most complicated ones) are not supported. But more generally, this means that if we have `Concat` followed by `ConstantOfShape` in our graph, and `Concat` involves non-const tensors, then we're out of luck.

I don't think it is impossible to make it work, but I also don't see a simple solution. I was hoping for one, though :(

**Comment by jfix71** [comment]
>Well, this effectively means that some graphs produced by `torch.onnx` for some models (and not the most complicated ones) are not supported.

This is somewhat expected -- we started with core functionality and have expanded that functionality over time. In addition to lack of dynamic shapes, Glow also doesn't support control flow (e.g. for RNNs we currently must pre-unroll them to support them). If you're interested in driving forward dynamic shape support we'd be happy to discuss different designs and welcome any contributions.

**Comment by wyessen** [comment]
@jfix71 Thanks! Yeah, I'll probably work this, along with the other issues I've raised. I'm currently working on two major additions (a generic model runner referenced in the other issues, and a generic inference engine that is able to either use static bundles or link them in dynamically). These are already here: https://github.com/wyessen/glow/commit/5de9e626da29b4d35127cb1b385a263f213e9f6c#diff-d088ac472468cc7f5a0039b7508a691a, but are not quite ready for a PR... stay tuned.
